### PR TITLE
Implement FindMeetingQuery to find all available meeting time ranges for a set of attendees, given existing events.

### PR DESCRIPTION
**Pull request by KathyFeiyang**
Implement FindMeetingQuery to find all available meeting time ranges for a set of attendees, given existing events [Pass the provided tests]:
- Handle edge cases, such as when there are no attendees, or existing events.
- Find the available time ranges by examining the available time between (1) an ending time range occupied by an existing event & (2) the immediately following, starting time range occupied by another existing event.
- Check two additional time ranges: one occurring before the first existing event, and one occurring after the last existing event.

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,120 @@
...
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> availableTimeRanges = new ArrayList<>();
+    Set<String> attendees = new HashSet<>(request.getAttendees());
+    long duration = request.getDuration();
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: rename this to requestedDuration

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,120 @@
...
+
+    if (Long.compare(duration, TimeRange.WHOLE_DAY.duration()) > 0) {
+      // If the requested duration is longer than the {@code WHOLE_DAY}, there can't be any
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Remove the comments inside this if/else block. They're not adding anything useful that your code doesn't already make clear

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,120 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:29]: Add a function description for query

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,120 @@
...
+      // If there are no existing {@code Event} or mandatory attendees, the {@code WHOLE_DAY}
+      // satisfies the request.
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // Find existing {@code TimeRange} that correspond to {@code Event} having overlapping
+    // attendees with the {@code request}.
+    Collection<TimeRange> relevantTimeRanges = new HashSet<>();
+    for (Event event : events) {
+      // Make a shallow copy of the original set of attendees.
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Remove this comment

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,120 @@
...
+      // If the requested duration is longer than the {@code WHOLE_DAY}, there can't be any
+      // {@code TimeRange} that satisfies the {@code request}.
+      return Arrays.asList();
+    } else if (events.isEmpty() || attendees.isEmpty()) {
+      // If there are no existing {@code Event} or mandatory attendees, the {@code WHOLE_DAY}
+      // satisfies the request.
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // Find existing {@code TimeRange} that correspond to {@code Event} having overlapping
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This should be made part of the function description.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,120 @@
...
+    for (Event event : events) {
+      // Make a shallow copy of the original set of attendees.
+      Set<String> overlappingAttendees = new HashSet<>(event.getAttendees());
+      overlappingAttendees.retainAll(attendees);
+      if (!overlappingAttendees.isEmpty()) {
+        relevantTimeRanges.add(event.getWhen());
+      }
+    }
+    if (relevantTimeRanges.isEmpty()) {
+      // If there are no existing {@code Event} that contain overlapping attendees with the meeting
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Remove this comment. Instead of explaining **what** your code does, try to explain **why** it does what it does. Another programmer reading your code can figure out what it does, but it's not so obvious why it is behaving this way. Try to make the function assumptions, contracts etc. more explicit in the function description itself.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,120 @@
...
+      // request, the {@code WHOLE_DAY} satisfies the {@code request}.
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // Sort the relevant {@code TimeRange} based on their start and end times respectively.
+    List<TimeRange> startOrderedTimeRanges = new ArrayList<>(relevantTimeRanges);
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Also, this function is getting too big. Split it up :)

**Review by suparnabarua — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 6 above)
Comment (KathyFeiyang) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Got it, thank you!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -14,10 +14,169 @@
...
+                                          availableTimeRanges);
+
+    // Find available {@code TimeRange} by looking at the available time between one ending
+    // {@code TimeRange} and the immediately following starting {@code TimeRange}.
+    ListIterator<TimeRange> startOrderedTimeRangesIterator = startOrderedTimeRanges.listIterator();
+    ListIterator<TimeRange> endOrderedTimeRangesIterator = endOrderedTimeRanges.listIterator();
+    TimeRange endingTimeRange = endOrderedTimeRangesIterator.next();
+    TimeRange startingTimeRange = startOrderedTimeRangesIterator.next();
+    while (true) {
+      // There is no available time after the current ending {@code TimeRange} and before the
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Remove this comment. Try to avoid over documenting your code and only add comments for WHY certain parts exist, not WHAT they do

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -14,10 +14,169 @@
...
+                                                                TimeRange.ORDER_BY_END);
+
+    // Potentially add the {@code TimeRange} before the first existing {@code TimeRange}.
+    TimeRange firstStartingTimeRange = startOrderedTimeRanges.get(0);
+    checkDurationAndAddAvailableTimeRange(requestedDuration,
+                                          TimeRange.START_OF_DAY, firstStartingTimeRange.start(),
+                                          availableTimeRanges);
+
+    // Find available {@code TimeRange} by looking at the available time between one ending
+    // {@code TimeRange} and the immediately following starting {@code TimeRange}.
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Reword this comment. The " and the immediately following starting {@code TimeRange}" doesn't really make sense.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -14,10 +14,169 @@
...
+      if (endingTimeRange.overlaps(startingTimeRange)) {
+        if (startOrderedTimeRangesIterator.hasNext()) {
+          startingTimeRange = startOrderedTimeRangesIterator.next();
+        } else {
+          break;
+        }
+        continue;
+      }
+      // There is another {@code TimeRange} that ends later than the current ending
+      // {@code TimeRange} + before the immediately following, starting {@code TimeRange}.
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I see you mention "before the immediately following, starting" a few times, but I'm not sure what you mean? Do you mean before the next available starting Timerange?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -14,10 +14,169 @@
...
+  * Finds the existing {@code TimeRange} that correspond to {@code Event} having
+  * overlapping attendees with a set of attendees that we are concerned with.
+  */
+  private Collection<TimeRange> getRelevantTimeRangesFromEvents(Collection<Event> events,
+      Set<String> attendees) {
+    Collection<TimeRange> relevantTimeRanges = new HashSet<>();
+    for (Event event : events) {
+      Set<String> overlappingAttendees = new HashSet<>(event.getAttendees());
+      overlappingAttendees.retainAll(attendees);
+      if (!overlappingAttendees.isEmpty()) {
```
Comment (kellyktan) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: You could consider using https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#disjoint(java.util.Collection,%20java.util.Collection)

Comment (kellyktan) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: You could consider using https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#disjoint(java.util.Collection,%20java.util.Collection) so that you don't have to make additional copies of the read-only attendees.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 12]
```diff
@@ -14,10 +14,169 @@
...
+  * Then finds the available {@code TimeRange} sandwiched between an existing, ending
+  * {@code TimeRange} and an existing starting {@code TimeRange}.
+  *
+  * @param events Existing events that each occupy some {@code TimeRange}.
+  * @param request A meeting request for some mandatory and optional attendees and a duration.
+  * @return All available {@code TimeRange} that satisfy the meeting request.
+  */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Set<String> attendees = new HashSet<>(request.getAttendees());
```
Comment (kellyktan) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: It doesn't look like you need to copy this as a non-read-only set.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 13]
```diff
@@ -14,10 +14,169 @@
...
+                                          availableTimeRanges);
+
+    return availableTimeRanges;
+  }
+
+  /**
+  * Finds the existing {@code TimeRange} that correspond to {@code Event} having
+  * overlapping attendees with a set of attendees that we are concerned with.
+  */
+  private Collection<TimeRange> getRelevantTimeRangesFromEvents(Collection<Event> events,
```
Comment (kellyktan) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: I would move this above getAvailabeTimeRanges() since you're using it first.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 14]
```diff
@@ -14,10 +14,169 @@
...
+      overlappingAttendees.retainAll(attendees);
+      if (!overlappingAttendees.isEmpty()) {
+        relevantTimeRanges.add(event.getWhen());
+      }
+    }
+    return relevantTimeRanges;
+  }
+
+  /**
+  * Sorts a collection of {@code TimeRange} based on the specified {@code sortOrder}.
```
Comment (kellyktan) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: "Sorts {@code timeRanges}" since it's a param in this function.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 13 above)
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: You should change the name of this function. What makes a TimeRange relevant? Perhaps getOverlappingTimeRanges?

**Review by suparnabarua — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 15]
```diff
@@ -14,10 +14,169 @@
...
+    return orderedTimeRanges;
+  }
+
+  /**
+  * If a [start, end) represented {@code TimeRange} has a sufficient duration, adds it to the
+  * set of available {@code TimeRange}.
+  */
+  private void checkDurationAndAddAvailableTimeRange(long requestedDuration, int start, int end,
+      Collection<TimeRange> availableTimeRanges) {
+    if (Long.compare(requestedDuration, end - start) <= 0) {
```
Comment (kellyktan) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Since requestDuration, end, and start are all primitive types, I don't think you need to use the Long class to compare them (i.e. can just do requestedDuration <= end - start)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 16]
```diff
@@ -14,10 +14,169 @@
...
+                                          TimeRange.START_OF_DAY, firstStartingTimeRange.start(),
+                                          availableTimeRanges);
+
+    // Find available {@code TimeRange} by looking at the available time between one ending
+    // {@code TimeRange} and the immediately following starting {@code TimeRange}.
+    ListIterator<TimeRange> startOrderedTimeRangesIterator = startOrderedTimeRanges.listIterator();
+    ListIterator<TimeRange> endOrderedTimeRangesIterator = endOrderedTimeRanges.listIterator();
+    TimeRange endingTimeRange = endOrderedTimeRangesIterator.next();
+    TimeRange startingTimeRange = startOrderedTimeRangesIterator.next();
+    while (true) {
```
Comment (kellyktan) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:145]: Maybe consider using a
```
do {
  ...
} while (loop break condition);
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 17]
```diff
@@ -14,10 +14,169 @@
...
+      // There is no available time after the current ending {@code TimeRange} and before the
+      // immediately following starting {@code TimeRange}.
+      if (endingTimeRange.overlaps(startingTimeRange)) {
+        if (startOrderedTimeRangesIterator.hasNext()) {
+          startingTimeRange = startOrderedTimeRangesIterator.next();
+        } else {
+          break;
+        }
+        continue;
```
Comment (kellyktan) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: can just put this up in the second nested `if` after `startingTimeRange =` line since you're `break`ing in the `else` anyways.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 18]
```diff
@@ -14,10 +14,169 @@
...
+      }
+      // There is another {@code TimeRange} that ends later than the current ending
+      // {@code TimeRange} + before the immediately following, starting {@code TimeRange}.
+      // (Since there exists a {@code TimeRange} starting later than the current ending
+      //  {@code TimeRange}, there must be other {@code TimeRange} ending even later. Therefore,
+      //  the {@code endOrderedPointer} is guaranteed to not go out of range.)
```
Comment (kellyktan) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: extra spaces?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 10 above)
Comment (KathyFeiyang) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Yes

Comment (KathyFeiyang) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I'll reword the comments. Thanks for pointing this out!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 16 above)
Comment (kellyktan) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:145]: Although, there are so many break/continue points that this may not be the cleanest...

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 13 above)
Comment (KathyFeiyang) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Thanks for the feedback! I modified it to getConcernedAttendeesTimeRangesFromEvents

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 16 above)
Comment (kellyktan) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:145]: Maybe consider instead of the "continues" it might be cleaner to use nested while loops? As is, the logic in this function is a bit hard to follow (although not sure if any of my suggestions would make it easier...).  At least it looks like it works, so feel free to ignore my suggestions here unless you feel inspired lol.

**Review by kellyktan — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 17 above)
Comment (KathyFeiyang) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Yeah, I've considered this. The reason why I didn't do it is because I wanted to keep the logic for the iterator (the if-block) separate from the rest. It's basically pointer++, but it looks long. I'll try to make this easier to read

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 16 above)
Comment (KathyFeiyang) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:145]: I modified the code to include nested while loops rather than `continue`s. Hopefully the code has improved readability. I feel that portions of the confusion comes from the bound-checking template code for iterators:
`      if (iterator.hasNext()) {
        item = iterator.next();
      } else {
        break;}`
I might switch to arrays (which allows efficient random access) and manipulate pointers instead.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 19]
```diff
@@ -54,15 +54,31 @@
...
     Collection<TimeRange> occupiedTimeRanges =
-        getRelevantTimeRangesFromEvents(events, attendees);
+        getConcernedAttendeesTimeRangesFromEvents(events, attendees);
 
     return getAvailabeTimeRanges(occupiedTimeRanges, requestedDuration);
   }
 
+  /**
+  * Finds the existing {@code TimeRange} that correspond to {@code Event} having
+  * overlapping attendees with a set of attendees that we are concerned with.
```
Comment (kellyktan) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: use param names when available (i.e. {@code events}, {@code attendees})

Comment (kellyktan) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I would also change this to be more about "what" you're doing rather than "how" (i.e. instead of "overlapping attendees", maybe something like "timeranges of events that include the given attendees")

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 20]
```diff
@@ -76,41 +92,45 @@
...
-        continue;
       }
-      // There is another {@code TimeRange} that ends later than the current ending
-      // {@code TimeRange} + before the immediately following, starting {@code TimeRange}.
-      // (Since there exists a {@code TimeRange} starting later than the current ending
-      //  {@code TimeRange}, there must be other {@code TimeRange} ending even later. Therefore,
-      //  the {@code endOrderedPointer} is guaranteed to not go out of range.)
+
+      // There is another occupied {@code TimeRange} between the current ending {@code TimeRange}
+      // and the next occupied {@code TimeRange}.
```
Comment (kellyktan) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Maybe instead, something like "Find the latest occupied timerange before startingtimerange".

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 21]
```diff
@@ -76,41 +92,45 @@
...
+      // (Since there exists a {@code TimeRange} that occurs later than the current ending
+      // {@code TimeRange}, we can be certain that invoking .next() won't go out of range.)
       TimeRange nextEndingTimeRange = endOrderedTimeRangesIterator.next();
```
Comment (kellyktan) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I feel like you don't necessarily need a new variable for "next".  Given that iterator.previous() will still return the previous element, you can probably do this directly overwriting endingtimerange

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 16 above)
Comment (kellyktan) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:145]: Or, instead of using multiple levels of breaks, you could refactor this out to a helper function and just "return" at the places you do `break findGapsInOccupiedTimeRanges`.

**Review by kellyktan — APPROVED** [review]
Some nits and optional suggestions.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 21 above)
Comment (KathyFeiyang) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Yes, it's not necessary. I originally intended to use "nextEndingTimeRange" to describe the logic of the code. I'll make the change

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 22]
```diff
@@ -14,10 +14,182 @@
...
+  *   |----ending {@code TimeRange}----| available time |----starting {@code TimeRange}----|
+  *
+  * @param requestedDuration The requested length of the overlapping, available {@code TimeRange}
+  *   to find.
+  * @param startOrderedTimeRanges A start-time ordered list of occupied {@TimeRange}.
+  * @param endOrderedTimeRanges An end-time ordered list of occupied {@TimeRange}.
+  * @param availableTimeRanges A list of available {@code TimeRange} that avoid the already
+  *   occupied {@code TimeRange}.
+  */
+  private void findAndAddGapsInOccupiedTimeRanges(long requestedDuration,
```
Comment (kellyktan) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:136]: nit: first param should be on newline and aligned with other params?

**Review by kellyktan — APPROVED** [review]
LGTM!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 23]
```diff
@@ -14,10 +14,185 @@
...
 public final class FindMeetingQuery {
+
+  /**
+   * Finds the available {@code TimeRange} for a set of attendees to satisfy a meeting
+   * {@code request}, given some existing {@code events} and their respective attendees.
+   * The resultant {@code TimeRange} do not overlap with one another and cover all possible
+   * empty gaps in the existing {@code events}. These {@code TimeRange} have the same length as
+   * or are longer than the requested meeting duration.
+   * -- First extracts the existing {@code TimeRange} of those {@code Event} that include any
+   * attendees mentioned in the {@code request}; we only need to avoid conflicting with
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Avoid semicolons. Either start a new sentence or use conjunctions

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 24]
```diff
@@ -14,10 +14,185 @@
...
+
+    if (requestedDuration > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    } else if (events.isEmpty() || attendees.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collection<TimeRange> occupiedTimeRanges =
+        getConcernedAttendeesTimeRangesFromEvents(events, attendees);
+
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Remove the extra new line here. While it makes sense to add new lines between large blocks of code, that's unnecessary here.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 25]
```diff
@@ -14,10 +14,185 @@
...
+    List<TimeRange> endOrderedTimeRanges = getOrderedTimeRanges(occupiedTimeRanges,
+                                                                TimeRange.ORDER_BY_END);
+    List<TimeRange> availableTimeRanges = new LinkedList<>();
+
+    // Potentially add the {@code TimeRange} before the first occupied {@code TimeRange}.
+    TimeRange firstStartingTimeRange = startOrderedTimeRanges.get(0);
+    checkDurationAndAddAvailableTimeRange(requestedDuration,
+                                          TimeRange.START_OF_DAY, firstStartingTimeRange.start(),
+                                          availableTimeRanges);
+
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Remove the new line here

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 26]
```diff
@@ -14,10 +14,185 @@
...
+    findAndAddGapsInOccupiedTimeRanges(requestedDuration,
+                                       startOrderedTimeRanges, endOrderedTimeRanges,
+                                       availableTimeRanges);
+
+    // Potentially add the {@code TimeRange} after the last occupied {@code TimeRange}.
+    TimeRange lastEndingTimeRange = endOrderedTimeRanges.get(endOrderedTimeRanges.size() - 1);
+    checkDurationAndAddAvailableTimeRange(requestedDuration,
+                                          lastEndingTimeRange.end(), TimeRange.END_OF_DAY + 1,
+                                          availableTimeRanges);
+
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Remove the new line here

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 16 above)
Comment (suparnabarua) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:145]: Maybe you should switch to using ArrayLists to make accessing the elements clearer?

> reply:
I will use arrays instead for O(1) random access and simplicity of the data structure. However ArrayLists also provide O(1) random access and would also be an option

**Review by suparnabarua — CHANGES_REQUESTED** [review]

**Review by suparnabarua — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 16 above)
Comment (kellyktan) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:145]: It still seems a bit confusing that you both have variables for the pointers plus the actual timerange object (i.e. that they can be updated independently, potentially). Maybe instead of having the timerange object, just using the orderedTimeRanges[ptr] might be better? +suparnabarua@google.com what do you think?

Comment (KathyFeiyang) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:145]: I have that code ready now. I personally think that the code does look clearer

Comment (suparnabarua) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:145]: yeah that's a good point about using the pointers instead!
